File: script/bun/src/puzzle/beginner/q07.ts
/**
 * YYYYMMDD形式の日付を、2進数→逆にする→10進数としたとき、元の値と同じになるものは？
 */

import { dayjs } from '~/lib/dayjs.js';

const begin = performance.now();

const FROM = dayjs('1964-10-10');
const END = dayjs('2020-07-24');

for (let i = FROM; i.isSameOrBefore(END); i = i.add(1, 'day')) {
  const base = Number.parseInt(i.format('YYYYMMDD'));
  const b02 = base.toString(2).split('').reverse().join('');
  const transformed = Number.parseInt(b02, 2);

  if (base === transformed) {
    console.log(`a07. = ${i.format('YYYY-MM-DD')}`);
  }
}

const finish = performance.now();
console.log(`${finish - begin}ms`);

File: script/java/app/src/main/java/org/example/puzzle/beginner/Q07.java
package org.example.puzzle.beginner;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public class Q07 {
  public static void main(String[] args) {
    final var begin = System.currentTimeMillis();

    final LocalDate FROM = LocalDate.of(1964, 10, 10);
    final LocalDate END = LocalDate.of(2020, 7, 24);

    for (LocalDate i = FROM; i.isBefore(END) || i.isEqual(END); i = i.plusDays(1)) {
      final var base = Integer.parseInt(i.format(DateTimeFormatter.ofPattern("yyyyMMdd")));
      final var b02 = new StringBuilder(Integer.toString(base, 2)).reverse().toString();
      final var transformed = Integer.parseInt(b02, 2);

      if (base == transformed) {
        System.out.println("a07. = " + i.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")));
      }
    }

    final var finish = System.currentTimeMillis();
    System.out.println((finish - begin) + "ms");
  }
}
